lib/1.4/dml-builtins.dml

```diff
@@ -1058,19 +1067,55 @@ template port is (object, _qname) {
     param limitations = undefined;
     param obj = (simics_api_version == "5" || simics_api_version == "4.8")
         #? dev.obj #: _port_obj();
+    param _is_subdevice_port : bool;
+    param _is_subdevice_port default false;
+    #if (this._nongroup_parent.objtype == "port"
+         #? !this._nongroup_parent._is_subdevice_port #: false) {
+        error "Nested ports are only permitted if the outer port instantiates"
+            + " the subdevice_port template";
+    }
+
     session conf_object_t *_cached_port_obj;
     shared method _port_obj() -> (conf_object_t *) {
         if (this._cached_port_obj != NULL) {
             return this._cached_port_obj;
         }
-        local char name[strlen(this._qname()) + 6];
-        strcpy(name, "port.");
-        strcpy(name + 5, this._qname());
+        local const char *qn = this._qname();
+        local int len = strlen(qn);
+        local int anc_len;
+        for (anc_len = strlen(qn); anc_len > 0; --anc_len) {
+            if (qn[anc_len - 1] == '.')
+                break;
+        }
+        local char name[len + 6];
+        memcpy(name, qn, anc_len);
+        memcpy(name + anc_len, "port.", 5);
+        strcpy(name + anc_len + 5, qn + anc_len);
         this._cached_port_obj = SIM_object_descendant(dev.obj, name);
         return this._cached_port_obj;
     }
 }
 
+/**
+## Port templates
+
+There is a single template for [`port`](#port-objects) objects, namely
+`subdevice_port`. The template signifies that the port object primarily
+represents a logical subsystem of the device, rather than an incoming signal.
+This has the effect that the name of the corresponding Simics configuration
+object will not include the implicit `.port` prefix. A port that implements
+this template may not be named `port` or `bank`, as that potentially can cause
+name clashes.
+*/
+template subdevice_port is port {
+    param _is_subdevice_port = true;
+    // this prevents name collisions on port objects
+    #if (this.name == "port" || this.name == "bank") {
+         error "A port that instantiates subdevice_port may not be"
+             + " named 'port' or 'bank'";
+    }
+}
```

Comment:
Nothing in DMLC or in builtins prevents the `.port` prefix from being used when genning classes for ports instantiating this template.
So for the following code:
```
port p is subdevice_port {
  port c;
}
```
We'll get the objects `dev.port.p` and `dev.p.port.c`. Is that really what we want?

Reply:
no, subdevice_port suppresses the extra .port (see [group_hierarchy.dml](66/files#diff-d298a95a1a3810a2e160680baae31a4877945cfbd74f12ba7649ab7d96d53e8d)), so the objects are named `dev.p` and `dev.p.port.c`.

Reply:
It suppresses the _extra_ port for `the `c` child port, but _not_ for `p`. To confirm, I added this to the `group_hierarchy.dml` tests:
```
    local conf_object_t *ol = SIM_object_descendant(
        dev.obj, "port.pa[1][2]");
    assert ol != NULL;
```
And indeed, the test still passed.

Reply:
Right, this is a bug in `c_backend.py` (now fixed locally). Another bug I found is that for nested ports, the inner port is registered on the device instead of the outer port; i.e. my new test `assert SIM_port_object_parent(port.pa[1][2].p[4][6].obj) == port.pa[1][2].obj` fails.